private/getParProfilesByOptimization.m

```diff
@@ -246,24 +249,44 @@
                 abs(dtheta(iPar)),options.options_getNextPoint.min,options.options_getNextPoint.max,options.options_getNextPoint.update,...
                 -(log(1-options.dR_max)+options.dJ*(logPost-logPost_max)+logPost),negLogPost,...
-                parameters.constraints,options.options_getNextPoint.mode,iPar);
+                parameters.constraints,options.options_getNextPoint.mode,iPar, options.localOptimizer);
             negLogPostReduced = setObjectiveWrapper(objective_function, options, 'negative log-posterior', iPar, theta_next(iPar), true, true);
             
+            % Check, if Hessian should be used and if a Hessian function was set,
+            % otherwise use the third output of the objective function instead
+            if (strcmp(options.profileOptimizationOptions.Hessian, 'on'))
+                if (~isfield(options.profileOptimizationOptions, 'HessFcn') ...
+                        || isempty(options.profileOptimizationOptions.HessFcn))
+                    % this only works for box-constraints at the moment
+                    options.profileOptimizationOptions.HessFcn = @(varargin) HessianWrap(negLogPostReduced, varargin);
+                end
+            end
+            
             % Construction of reduced linear constraints
             [A,b,Aeq,beq] = getConstraints(theta,parameters,I);
             
             % Optimization
             try
-                [theta_I_opt,J_opt] = ...
-                    fmincon(negLogPostReduced,... % negative log-posterior function
-                    theta_next(I),...
-                    A  ,b  ,... % linear inequality constraints
-                    Aeq,beq,... % linear equality constraints
-                    parameters.min(I),...   % lower bound
-                    parameters.max(I),...   % upper bound
-                    [],options.profileReoptimizationOptions);    % options
+                switch options.localOptimizer
+                    case 'fmincon'
+                        [theta_I_opt,J_opt] = ...
+                            fmincon(negLogPostReduced,... % negative log-posterior function
+                            theta_next(I),...
+                            A  ,b  ,... % linear inequality constraints
+                            Aeq,beq,... % linear equality constraints
+                            parameters.min(I),...   % lower bound
+                            parameters.max(I),...   % upper bound
+                            [],options.profileOptimizationOptions);    % options
+                    case 'lsqnonlin'
+                        [theta_I_opt] = lsqnonlin(negLogPostReduced,...
+                            theta_next(I), ...
+                            parameters.min(I),...   % lower bound
+                            parameters.max(I),...   % upper bound
+                            options.profileOptimizationOptions);
+                        [~, ~, J_opt] = negLogPostReduced(theta_I_opt);
```

Comment:
I am pretty sure we can use the output of `lsqnonlin` to save us this call to `negLogPostReduced`

Reply:
Well... I'm not so sure about that...
The point is that using lsqnonlin and estimating the standard deviation forces you to add a constant term to the likelihood. This constant term may depend on your implementation (unless we require the user to stick to a specific form).
As Pesto doesn't now this constant term, there's no way of getting it out of lsqnonlin, which would mean that the likelihood values from lsqnonlin and fmincon are not directly comparable...
This call to negLogPostReduced is there for computing the likelihood without this offset...
If anyone has a better idea how to deal with this problem, it would be very welcome.

Reply:
I would probably also remove the call. If I understand correctly, it requires the used to implement two functions. I would rather provide a warning to inform the user that he/she has to handle it outside.

Reply:
No, the user has to provide only one cost function. It has the following ways of being called:
[residuals] = objectiveFunction(...)
[residuals, Jacobian] = objectiveFunction(...)
[residuals, ~, log-likelihood] = objectiveFunction(...)

Reply:
Ah, okay. Thanks for the clarification.

Reply:
Ehm, isn't this really messing things up when you use lsqnonlin for profile calculation? Using the same name for different things (i.e. the same as internal values in optimization vs different) is asking for trouble. If the objective function and the log-likelihood do not match indicate this by using respective variable names.

Reply:
Hm, the use of different names would require changes in many other routines, e.g. visualization routines.

Reply:
I would guess that most of these routines currently assume that the objective function == the likelihood function, so they will probably need updates anyways. For the visualization routines this is a matter of copy&paste, so this should not be an issue.

I am primarily worried about functions like `optimizeProfileForParameterI`, where this might easily get mixed up.

Reply:
I opened up #152 for this.
The point is: The implementation right now works for getMultiStarts and getParameterProfiles. getPropertyProfiles() will be complicated anyway, as this uses nonlinear equality constraints, of which I'm not sure whether this is possible with lsqnonlin...

Reply:
Any votes whether I should change the implementation for the current PR?

Reply:
I am not so much worried about the implementation of the likelihood function. If the function signature is well documented, this should be fine. What I am worried about is the way the values are stored after optimization. 
If something is called J_opt, I assume this is equal to the objective function value that is used in the optimization. If I use J_opt as reference point for my profile calculation and directly use the result from the optimizer, I will be screwed and calculate wrong profiles, as I am missing the offset.
Also, creating different variables for J and nllh will make the code more transparent and reasonable, as it should become evident why we need to call the likelihood function again (you actually don't, you can compute the offset once and add it after every optimization).

Reply:
It is consistent between profiling and optimization right now. So you can use either lsqnonlin or fmincon for optimization and profiling and switch between those two. The only thing you need to make sure when switching is to provide the correct objective function.

Reply:
I like having always the log-posterior stored somewhere. Converting values manually can be rather cumbersome at some point. Maybe, we could simply include for the use of lsqnonlin two fields, e.g. neglogPost_opt and chi2_opt. To ensure consistency, if residuals are used, both are provided but only neglogPost_opt is used for downstream stuff.

In this case, we should also state neglogPost_opt and chi2_opt instead of J_opt. In the profile calculation routines, we could simply used neglogPost_opt for the stepping, ... by specifying the interface.

Reply:
Well, right now, negLogPost is stored always. In optimization and profiling.
It would nevertheless make sense to also store chi2 for Lsqnonlin. I pushed a version with a better naming today. However, is there something on which we can agree?
Would everyone be fine with e.g. always storing the log-likelihood? And computing the offset once and using it later on?